code/datums/status_effects/debuffs/fire_stacks.dm

```diff
@@ -0,0 +1,263 @@
+	var/datum/weakref/firelight_ref
+	/// Type of mob light emitter we use when on fire
+	var/firelight_type = /obj/effect/dummy/lighting_obj/moblight/fire
+	/// Stores current fire overlay icon state, for optimisation purposes
+	var/last_icon_state
+
+/datum/status_effect/fire_handler/fire_stacks/tick(delta_time, times_fired)
+	if(stacks <= 0)
+		qdel(src)
+		return TRUE
+
+	if(!on_fire || isanimal(owner))
+		return TRUE
+
+	if(iscyborg(owner))
+		adjust_stacks(-0.55 * delta_time)
+	else
+		adjust_stacks(-0.05 * delta_time)
+
+	if(stacks <= 0)
+		qdel(src)
+		return TRUE
+
+	var/datum/gas_mixture/air = owner.loc.return_air()
+	if(!air.gases[/datum/gas/oxygen] || air.gases[/datum/gas/oxygen][MOLES] < 1)
+		qdel(src)
+		return TRUE
+
+	deal_damage(delta_time, times_fired)
+	update_overlay()
+
+/**
+ * Proc that handles damage dealing and all special effects
+ *
+ * Arguments:
+ * - delta_time
+ * - times_fired
+ *
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/deal_damage(delta_time, times_fired)
+	owner.on_fire_stack(delta_time, times_fired, src)
+
+	var/turf/location = get_turf(owner)
+	location.hotspot_expose(700, 25 * delta_time, TRUE)
+
+/**
+ * Used to deal damage to humans and count their protection.
+ *
+ * Arguments:
+ * - delta_time
+ * - times_fired
+ * - no_protection: When set to TRUE, fire will ignore any possible fire protection
+ *
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/harm_human(delta_time, times_fired, no_protection = FALSE)
+	var/mob/living/carbon/human/victim = owner
+	var/thermal_protection = victim.get_thermal_protection()
+
+	if(thermal_protection >= FIRE_IMMUNITY_MAX_TEMP_PROTECT && !no_protection)
+		return
+
+	if(thermal_protection >= FIRE_SUIT_MAX_TEMP_PROTECT && !no_protection)
+		victim.adjust_bodytemperature(5.5 * delta_time)
+		return
+
+	victim.adjust_bodytemperature((BODYTEMP_HEATING_MAX + (stacks * 12)) * 0.5 * delta_time)
+	SEND_SIGNAL(victim, COMSIG_ADD_MOOD_EVENT, "on_fire", /datum/mood_event/on_fire)
+	victim.mind?.add_memory(MEMORY_FIRE, list(DETAIL_PROTAGONIST = victim), story_value = STORY_VALUE_OKAY)
+
+/**
+ * Handles mob ignition, should be the only way to set on_fire to TRUE
+ *
+ * Arguments:
+ * - silent: When set to TRUE, no message is displayed
+ *
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/ignite(silent = FALSE)
+	if(HAS_TRAIT(owner, TRAIT_NOFIRE))
+		return FALSE
+
+	on_fire = TRUE
+	if(!silent)
+		owner.visible_message(span_warning("[owner] catches fire!"), span_userdanger("You're set on fire!"))
+
+	if(firelight_type)
+		firelight_ref = WEAKREF(new firelight_type(owner))
+
+	SEND_SIGNAL(owner, COMSIG_LIVING_IGNITED, owner)
+	cache_stacks()
+	update_overlay()
+
+/**
+ * Handles mob extinguishing, should be the only way to set on_fire to FALSE
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/extinguish()
+	if(firelight_ref)
+		var/atom/firelight = firelight_ref.resolve()
+		if(firelight && !QDELETED(firelight))
+			qdel(firelight)
```

Comment:
You can just qdel the reference and it'll do this for ya 
![image](https://user-images.githubusercontent.com/58055496/166422284-6dc3891a-95d2-44ea-a194-564c46dc8719.png)